README.md
# 獅尾蝙蝠圓體 Swei Bat Sans

獅尾蝙蝠圓體基於[思源黑體](https://github.com/adobe-fonts/source-han-sans)筆觸長耳朵和拔腳改造，更加簡明現代化的字體。支援简体中文、繁體中文、韓文與日文；可以免費商用，歡迎大家自由應用、自由優化、自由改作！

獅尾三角黑體的特色是，「一」字的開頭和結尾筆觸由平頭換成蝙蝠耳朵形狀。

![字體預覽](https://github.com/max32002/swei-bat-sans/raw/master/preview/welcome.png)

和思源黑體一樣，支援7種的字重：
![字體比較預覽](https://github.com/max32002/swei-bat-sans/raw/master/preview/compare_style.png)
附註：目前有些筆畫還沒有套用到效果，是程式需要改進的地方。

## 與其他字體的比較
* 在「刂」的筆畫，獅尾比較相似台灣(教育部國字標準字體)教育部推薦字體筆順。
* 在「肉」、「糸」、「女」、「辶」、「食」的筆畫，獅尾比較相似台灣教育部推薦字體筆順，適合教育用途。
* 在「草」部的筆畫，獅尾是分開的二個部份。
* 字體裡的「草字頭部首」中間該不該連起來，「肉字旁部首」該不該變成「月」，要不要把手寫習慣代入印刷字體？整體看來，會不會影響視覺的延伸性？會不會影響印刷的可行性與閱讀的便利性？這個答案我也不清楚，如果你是台灣教育部標準字符的愛好者，獅尾字體應該是一個不錯的選擇。

## 更新日誌
[點擊此處](https://github.com/max32002/swei-bat-sans/blob/master/change_log.md) 查看更新記錄。

## 下載字型

請點選GitHub此畫面右上綠色「Clone or download」按鈕，並選擇「Download ZIP」，或點進想下載的ttf字型檔案，再點「Download」的按鈕進行下載。

## 網頁字型(Web Font)服務

網頁字型用於網頁上的字型顯示，訪客不需預先安裝字型檔，一樣能夠看到特殊的字型效果。不只是電腦，在智慧型手機和平板裝置的瀏覽器上也可正常顯示。實現該功能的原理是在瀏覽時才下載字型檔。

可以服用下面的css:
```
@font-face {
  font-family: 'SweiBatSansCJKtc-Regular';
  src: url(https://cdn.jsdelivr.net/gh/max32002/swei-bat-sans@2.0/WebFont/CJK%20TC/SweiBatSansCJKtc-Regular.woff2) format("woff2")
  , url(https://cdn.jsdelivr.net/gh/max32002/swei-bat-sans@2.0/WebFont/CJK%20TC/SweiBatSansCJKtc-Regular.woff) format("woff");
}
```

## 已知問題

* 這不是一個專業的字型檔案。
* 很多字還是可以看到思源黑體的的直角。
* 希望有贊助者，或其他勇者繼續改良和調整很多怪怪的小細節。
* 由於小編對字體編碼方式完全不清楚，相較於原版的思源黑體，可能有掉一些符號或不常用的字。

## 附註

* 演算黑科技將字體增加斜角改造，請參考 /python/ 目錄下的腳本檔案。透過調整程式碼，也許也可以產生出新的有趣字型。目前的程式應該還有很多錯誤或需要再加強的地方。請先把要處理的字型，轉成 FontForge 的 sfdir, 再透過程式去處理產生出來的檔案。

## 著作權與授權

* 本字型是基於 SIL Open Font License 1.1 改造Adobe和Google所開發、發表的「[思源黑體](https://github.com/adobe-fonts/source-han-sans)」字型。
* 本字型亦基於 SIL Open Font License 1.1 授權條款免費公開，關於授權合約的內容、免責事項等細節，請詳讀 License 文件。
    * 可自由商用 不需付費、知會或標明作者，即可自由使用此字型，亦可做商業應用。
    * 可自由傳布 可自由分享檔案、將檔案安裝於任何軟硬體中。
    * 可自由改作為其他字型 將字型檔案修改重製為其他字型檔案，改作後的字型檔案須同樣依 Open Font License 釋出。
    
    
## 相關網頁

花園家族：
* B2花園 B2 Hana
https://max-everyday.com/2020/08/b2-hana-font/
* 花園肉丸 Hana Meatball
https://max-everyday.com/2020/08/hana-meatball/

獅尾黑體家族：
* 獅尾霓黑體 Swei Bow Sans
https://max-everyday.com/2020/09/swei-bow-sans/
* 獅尾霓腿黑體 Swei Bow Leg
https://max-everyday.com/2020/09/swei-bow-leg/
* 獅尾蝙蝠圓體 Swei Bat Sans
https://max-everyday.com/2020/09/swei-bat-sans/
* 獅尾牙膏圓體 Swei Toothpaste
https://max-everyday.com/2020/09/swei-toothpaste/
* 獅尾三角黑體 Swei 3T Sans
https://max-everyday.com/2020/08/swei-3t-sans/
* 獅尾螺帽腿黑體 Swei Nut Leg
https://max-everyday.com/2020/08/swei-nut-leg/
* 獅尾螺帽黑體 Swei Nut Sans
https://max-everyday.com/2020/08/swei-nut-sans/
* 獅尾B2腿黑體 Swei B2 Leg
https://max-everyday.com/2020/07/swei-b2-leg/
* 獅尾B2黑體 Swei B2 Sans
https://max-everyday.com/2020/07/swei-b2-sans/
* 獅尾腿圓 Swei Gothic Leg
https://max-everyday.com/2020/08/swei-gothic-leg/
* 獅尾彩虹腿 Swei Rainbow Leg
https://max-everyday.com/2020/08/swei-rainbow-leg/
* 獅尾XD珍珠 Swei XD Pearl
https://max-everyday.com/2020/07/swei-xd-pearl/
* 獅尾D露西 Swei D Lucy
https://max-everyday.com/2020/07/swei-d-lucy/
* 獅尾半月字體 Swei Gothic
https://max-everyday.com/2020/04/swei-half-moon/
* 台灣圓體 TaiwanPearl
https://max-everyday.com/2020/06/taiwanpearl/
* 獅尾圓體 Swei Gothic
https://max-everyday.com/2020/04/swei-gothic/
* 獅尾黑體 Swei Sans
https://max-everyday.com/2020/03/swei-sans/

獅尾宋體家族：
* 獅尾B2宋朝 Swei B2 Serif
https://max-everyday.com/2020/07/swei-b2-serif/
* 獅尾肉丸 Swei Meatball
https://max-everyday.com/2020/06/swei-meatball/
* 獅尾四季春字體 Swei Spring
https://max-everyday.com/2020/04/swei-spring/

其他字體：
* 內海字體  Naikai Font
https://max-everyday.com/2020/03/naikaifont/
* 莫大毛筆字體 Bakudai Font
https://max-everyday.com/2020/03/bakudaifont/
* 正風毛筆字體 Masa Font
https://max-everyday.com/2020/05/masafont/
* 假粉圓體 Fake Pearl 
https://max-everyday.com/2020/03/open-huninn-font/
* 俊羽圓體 Yu Pearl 
https://max-everyday.com/2020/03/yupearl/

其他網站：
* 清松手寫體 JasonHandWriting
https://jasonfonts.max-everyday.com/
* Max學習字體相關的筆記
https://codereview.max-everyday.com/font-readme/

## 贊助Max

很高興可以替中華民國美學盡一分心力、讓台灣擁有更好的文字風景，希望能提供另一種美學讓大家選擇，如果你覺得這篇文章寫的很好，想打賞Max，贊助方式如下：
https://max-everyday.com/about/#donate


change_log.md
# 2020/09/05 ver 2.099
* 讓「蝙」字的戶部件套用到長耳朵的效果。原本的 ver 2.095版這一個部件沒套用到效果。

# 2020/09/04 ver 2.095
* 第一版。



python/util/Rule3_Water.py
#!/usr/bin/env python3
#encoding=utf-8

from . import spline_util
from . import Rule

# RULE # 3
# 處理三點「水」
# PS: 因為 array size change, so need redo.
class Rule(Rule.Rule):
    def __init__(self):
        pass

    def apply(self, spline_dict, resume_idx, inside_stroke_dict,apply_rule_log,generate_rule_log):
        redo_travel=False
        check_first_point = False

        MERGE_NEAR_POINTS_ACCURACY = 20

        # clone
        format_dict_array=[]
        format_dict_array = spline_dict['dots'][1:]
        format_dict_array = self.caculate_distance(format_dict_array)

        nodes_length = len(format_dict_array)
        #print("orig nodes_length:", len(spline_dict['dots']))
        #print("format nodes_length:", nodes_length)
        #print("resume_idx:", resume_idx)

        rule_need_lines = 4
        fail_code = -1
        if nodes_length >= rule_need_lines:
            for idx in range(nodes_length):
                if idx <= resume_idx:
                    # skip traveled nodes.
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                detect_code = format_dict_array[(idx+0)%nodes_length]['code']
                if detect_code in apply_rule_log:
                    if is_debug_mode:
                        print("match skip apply_rule_log +0:",detect_code)
                        pass
                    continue

                is_match_detect_code = False
                for detect_idx in range(4):
                    detect_code = format_dict_array[(idx+detect_idx)%nodes_length]['code']
                    if detect_code in generate_rule_log:
                        if is_debug_mode:
                            print("match skip generate_rule_log +%d:%s" % (detect_idx, detect_code))
                            pass
                if is_match_detect_code:
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                if is_debug_mode:
                    debug_coordinate_list = [[427,766]]
                    if not([format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y']] in debug_coordinate_list):
                        continue

                    print("="*30)
                    print("index:", idx)
                    for debug_idx in range(8):
                        print(debug_idx-2,": val#3:",format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['code'],'-(',format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['distance'],')')

                is_match_pattern = False

                # for:脚 811A「月」slide_percent_1: 0.84
                SLIDE_1_PERCENT_MIN = 0.50
                # for:蝙(uni8759)的戶，slide_percent_1: 1.81
                SLIDE_1_PERCENT_MAX = 1.86

                #TODO: 應該要檢查 SLIDE_1 + SLIDE_2 total max value.
                #  PS: 目前還沒有檢查。

                slide_percent_1 = spline_util.slide_percent(format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y'],format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'])
                slide_percent_2 = spline_util.slide_percent(format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'])
                if is_debug_mode:
                    print("slide_percent_1:",slide_percent_1)
                    print("slide_percent_2:",slide_percent_2)

                # 合併斜線：.uni811A 「脚」月上的斜線，分成二段。
                check_more_merge_stroke_condition = False
                if format_dict_array[(idx+0)%nodes_length]['distance'] > 150:
                    if format_dict_array[(idx+1)%nodes_length]['distance'] < 50:
                        if format_dict_array[(idx+2)%nodes_length]['distance'] < 50:
                            if format_dict_array[(idx+3)%nodes_length]['distance'] > 150:
                                if (format_dict_array[(idx+1)%nodes_length]['distance']+format_dict_array[(idx+2)%nodes_length]['distance']) >= 30:
                                    if (format_dict_array[(idx+1)%nodes_length]['distance']+format_dict_array[(idx+2)%nodes_length]['distance']) <= 120:
                                        check_more_merge_stroke_condition = True
                #print("check_more_merge_stroke_condition 1:", check_more_merge_stroke_condition)
                if check_more_merge_stroke_condition:
                    check_more_merge_stroke_condition = False
                    if format_dict_array[(idx+1)%nodes_length]['t'] == 'c':
                        if format_dict_array[(idx+2)%nodes_length]['t'] == 'c':
                            if format_dict_array[(idx+3)%nodes_length]['t'] == 'c':
                                check_more_merge_stroke_condition = True
                #print("check_more_merge_stroke_condition 2:", check_more_merge_stroke_condition)
                if check_more_merge_stroke_condition:
                    check_more_merge_stroke_condition = False
                    if format_dict_array[(idx+0)%nodes_length]['x_direction'] == -1 * format_dict_array[(idx+3)%nodes_length]['x_direction']:
                        if format_dict_array[(idx+1)%nodes_length]['x_direction'] == format_dict_array[(idx+0)%nodes_length]['x_direction']:
                            if format_dict_array[(idx+1)%nodes_length]['x_direction'] == format_dict_array[(idx+2)%nodes_length]['x_direction']:
                                check_more_merge_stroke_condition = True
                #print("check_more_merge_stroke_condition 3:", check_more_merge_stroke_condition)
                if check_more_merge_stroke_condition:
                    check_more_merge_stroke_condition = False
                    if format_dict_array[(idx+0)%nodes_length]['y_direction'] == -1 * format_dict_array[(idx+3)%nodes_length]['y_direction']:
                        if format_dict_array[(idx+1)%nodes_length]['y_direction'] == format_dict_array[(idx+3)%nodes_length]['y_direction']:
                            if format_dict_array[(idx+1)%nodes_length]['y_direction'] == format_dict_array[(idx+2)%nodes_length]['y_direction']:
                                check_more_merge_stroke_condition = True
                #print("check_more_merge_stroke_condition 4:", check_more_merge_stroke_condition)
                if check_more_merge_stroke_condition:
                    check_more_merge_stroke_condition = False
                    if slide_percent_1 >= SLIDE_1_PERCENT_MIN and slide_percent_1 <= SLIDE_1_PERCENT_MAX:
                        # for:脚 811A「月」slide_percent_2: 1.98
                        if slide_percent_2 >= 1.96:
                            check_more_merge_stroke_condition = True
                #print("check_more_merge_stroke_condition 5:", check_more_merge_stroke_condition)
                if check_more_merge_stroke_condition:
                    # start to merge line.
                    format_dict_array[(idx+1)%nodes_length]['distance'] = (format_dict_array[(idx+1)%nodes_length]['distance']+format_dict_array[(idx+2)%nodes_length]['distance'])

                    new_code = ' %d %d l 1\n' % (format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'])
                    format_dict_array[(idx+3)%nodes_length]['t'] = 'l'
                    format_dict_array[(idx+3)%nodes_length]['code'] = new_code
                    #print("new +3 code:", new_code)

                    del format_dict_array[(idx+2)%nodes_length]

                    if idx > (idx+2)%nodes_length:
                        idx -=1
                    nodes_length = len(format_dict_array)


                # 格式化例外：.31881 「閒」的上面的斜線。
                # convert ?cc? => ?cl?
                DISTANCE_IN_LINE_ACCURACY = 0.06
                if format_dict_array[(idx+1)%nodes_length]['t'] == 'c':
                    if format_dict_array[(idx+2)%nodes_length]['t'] == 'c':
                        if format_dict_array[(idx+1)%nodes_length]['match_stroke_width']:
                            if format_dict_array[(idx+0)%nodes_length]['x_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['x_direction']:
                                # 上邊或下邊，其中一條為接進水平線。
                                if format_dict_array[(idx+0)%nodes_length]['y_equal_fuzzy'] or format_dict_array[(idx+2)%nodes_length]['y_equal_fuzzy']:
                                    x1=format_dict_array[(idx+1)%nodes_length]['x']
                                    y1=format_dict_array[(idx+1)%nodes_length]['y']
                                    x2=format_dict_array[(idx+2)%nodes_length]['x']
                                    y2=format_dict_array[(idx+2)%nodes_length]['y']
                                    x2_1 = format_dict_array[(idx+2)%nodes_length]['x1']
                                    y2_1 = format_dict_array[(idx+2)%nodes_length]['y1']
                                    x2_2 = format_dict_array[(idx+2)%nodes_length]['x2']
                                    y2_2 = format_dict_array[(idx+2)%nodes_length]['y2']
                                    if spline_util.is_xyz_on_line(x1,y1,x2,y2,x2_1,y2_1,accuracy=DISTANCE_IN_LINE_ACCURACY) and spline_util.is_xyz_on_line(x1,y1,x2,y2,x2_2,y2_2,accuracy=DISTANCE_IN_LINE_ACCURACY):
                                        #print("convert c to l!")
                                        format_dict_array[(idx+2)%nodes_length]['x']=x2
                                        format_dict_array[(idx+2)%nodes_length]['y']=y2
                                        format_dict_array[(idx+2)%nodes_length]['t']='l'
                                        new_code = ' %d %d l 1\n' % (x2, y2)
                                        format_dict_array[(idx+2)%nodes_length]['code'] = new_code

                                        # [IMPORTANT] if change code, must triger check_first_point=True
                                        check_first_point=True

                # it's hard to handle, "complex" case in one time.
                # if let match case is fewer, it will match more condition.
                if format_dict_array[(idx+1)%nodes_length]['t'] == 'c':
                    if format_dict_array[(idx+2)%nodes_length]['t'] == 'l':
                        fail_code = 100
                        is_match_pattern = True

                # compare stroke_width
                if is_match_pattern:
                    fail_code = 200
                    #print(idx,"debug rule3:",format_dict_array[idx]['code'])
                    is_match_pattern = False
                    if format_dict_array[(idx+1)%nodes_length]['match_stroke_width']:
                        is_match_pattern = True

                # compare NEXT_DISTANCE_MIN
                if is_match_pattern:
                    fail_code = 210
                    is_match_pattern = False
                    if format_dict_array[(idx+0)%nodes_length]['distance'] > self.config.NEXT_DISTANCE_MIN:
                        is_match_pattern = True

                # check slide#1
                # 去除部份情況
                if is_match_pattern:
                    fail_code = 220
                    if slide_percent_1 < SLIDE_1_PERCENT_MIN:
                        is_match_pattern = False
                        if is_debug_mode:
                            print("slide_percent_1:",slide_percent_1)
                            print("SLIDE_1_PERCENT_MIN:",SLIDE_1_PERCENT_MIN)
                    if slide_percent_1 > SLIDE_1_PERCENT_MAX:
                        is_match_pattern = False
                        if is_debug_mode:
                            print("slide_percent_1:",slide_percent_1)
                            print("SLIDE_1_PERCENT_MAX:",SLIDE_1_PERCENT_MAX)

                # 做例外排除.
                # PS: 請不要判斷 (idx+2) and (idx+3), 因為超過矩形範圍。
                if is_match_pattern:
                    fail_code = 230
                    if format_dict_array[(idx+0)%nodes_length]['y_equal_fuzzy']:
                        if format_dict_array[(idx+1)%nodes_length]['y_equal_fuzzy']:
                            is_match_pattern = False
                    if format_dict_array[(idx+1)%nodes_length]['y_equal_fuzzy']:
                        if format_dict_array[(idx+2)%nodes_length]['y_equal_fuzzy']:
                            is_match_pattern = False
                    if format_dict_array[(idx+2)%nodes_length]['y_equal_fuzzy']:
                        if format_dict_array[(idx+3)%nodes_length]['y_equal_fuzzy']:
                            is_match_pattern = False
                    if format_dict_array[(idx+0)%nodes_length]['x_equal_fuzzy']:
                        if format_dict_array[(idx+1)%nodes_length]['x_equal_fuzzy']:
                            is_match_pattern = False
                    if format_dict_array[(idx+1)%nodes_length]['x_equal_fuzzy']:
                        if format_dict_array[(idx+2)%nodes_length]['x_equal_fuzzy']:
                            is_match_pattern = False

                # compare direction
                if is_match_pattern:
                    fail_code = 400
                    #print(idx,"debug rule3:",format_dict_array[idx]['code'])
                    is_match_pattern = False
                    if format_dict_array[(idx+0)%nodes_length]['x_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['x_direction']:
                        if format_dict_array[(idx+0)%nodes_length]['y_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['y_direction']:
                            is_match_pattern = True

                    # 追加例外：源 的下面小的水平腳
                    # 下面的 code, 直接讓 match=True, 需要小心處理！
                    # 二條「垂直線」。
                    if format_dict_array[(idx+0)%nodes_length]['x_equal_fuzzy']:
                        if format_dict_array[(idx+2)%nodes_length]['x_equal_fuzzy']:
                            # [重要]這個方向性，要相反! 
                            if format_dict_array[(idx+0)%nodes_length]['y_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['y_direction']:
                                is_match_pattern = True

                    # 二條「平行線」
                    if format_dict_array[(idx+0)%nodes_length]['y_equal_fuzzy']:
                        if format_dict_array[(idx+2)%nodes_length]['y_equal_fuzzy']:
                            # [重要]這個方向性，要相反! 
                            if format_dict_array[(idx+0)%nodes_length]['x_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['x_direction']:
                                is_match_pattern = True

                    # 追加例外：屑和備 的下面小的水平腳，由於無法套到 +0t x_equal_fuzzy, 也無法套到Rule#1.
                    # allow 
                    if format_dict_array[(idx+1)%nodes_length]['t']=='c':
                        if format_dict_array[(idx+2)%nodes_length]['y_equal_fuzzy']:
                            if format_dict_array[(idx+0)%nodes_length]['y'] > format_dict_array[(idx+1)%nodes_length]['y']:
                                # y axis 微距
                                if format_dict_array[(idx+0)%nodes_length]['y'] - format_dict_array[(idx+1)%nodes_length]['y'] <= 22:
                                    # 橫向的線。
                                    if format_dict_array[(idx+0)%nodes_length]['x_direction'] == -1 * format_dict_array[(idx+2)%nodes_length]['x_direction']:
                                        # 曲線往右下角長.
                                        # x1,y2 must between.
                                        if format_dict_array[(idx+1)%nodes_length]['x1'] >= format_dict_array[(idx+1)%nodes_length]['x'] :
                                            if format_dict_array[(idx+1)%nodes_length]['x2'] >= format_dict_array[(idx+1)%nodes_length]['x'] :
                                                if format_dict_array[(idx+1)%nodes_length]['x1'] <= format_dict_array[(idx+0)%nodes_length]['x'] :
                                                    if format_dict_array[(idx+1)%nodes_length]['x2'] <= format_dict_array[(idx+0)%nodes_length]['x'] :
                                                        # 力量要向下。
                                                        if format_dict_array[(idx+1)%nodes_length]['y1'] <= format_dict_array[(idx+0)%nodes_length]['y'] :
                                                            if format_dict_array[(idx+1)%nodes_length]['y2'] <= format_dict_array[(idx+0)%nodes_length]['y'] :
                                                                is_match_pattern = True


                # don't access the case for rule#1 & rule#2
                if is_match_pattern:
                    fail_code = 600
                    # must match horizonal or vertal equal.
                    if format_dict_array[(idx+1)%nodes_length]['x_equal_fuzzy']:
                        if format_dict_array[(idx+1)%nodes_length]['t']=='l':
                            is_match_pattern = False
                    if format_dict_array[(idx+1)%nodes_length]['y_equal_fuzzy']:
                        if format_dict_array[(idx+1)%nodes_length]['t']=='l':
                            is_match_pattern = False


                # check from bmp file.
                if is_match_pattern:
                    fail_code = 700
                    is_match_pattern = False

                    bmp_x1 = format_dict_array[(idx+0)%nodes_length]['x']
                    bmp_y1 = format_dict_array[(idx+0)%nodes_length]['y']
                    bmp_x2 = format_dict_array[(idx+1)%nodes_length]['x']
                    bmp_y2 = format_dict_array[(idx+1)%nodes_length]['y']
                    bmp_x3 = format_dict_array[(idx+2)%nodes_length]['x']
                    bmp_y3 = format_dict_array[(idx+2)%nodes_length]['y']
                    bmp_x4 = format_dict_array[(idx+3)%nodes_length]['x']
                    bmp_y4 = format_dict_array[(idx+3)%nodes_length]['y']
                    
                    # 精簡的比對，會出錯，例如「緫」、「縃」、「繳」的方。「解」的刀。「繫」的山。
                    #bmp_x2 = bmp_x1 + (5 * format_dict_array[(idx+0)%nodes_length]['x_direction'])
                    #bmp_x3 = bmp_x4 + (5 * format_dict_array[(idx+0)%nodes_length]['x_direction'])
                    #bmp_y2 = bmp_y1 + (5 * format_dict_array[(idx+0)%nodes_length]['y_direction'])
                    #bmp_y3 = bmp_y4 + (5 * format_dict_array[(idx+0)%nodes_length]['y_direction'])
                    
                    # full rectangle compare.
                    #inside_stroke_flag = self.is_inside_stroke(bmp_x1, bmp_y1, bmp_x2, bmp_y2, bmp_x3, bmp_y3, bmp_x4, bmp_y4)

                    # compact triangle compare, 
                    # allow one coner match to continue
                    inside_stroke_flag2 = False
                    inside_stroke_flag1,inside_stroke_dict = self.test_inside_coner(bmp_x1, bmp_y1, bmp_x2, bmp_y2, bmp_x3, bmp_y3, self.config.STROKE_MIN, inside_stroke_dict)
                    #print("inside_stroke_flag1:",inside_stroke_flag1)
                    if not inside_stroke_flag1:
                        # test next coner
                        inside_stroke_flag2,inside_stroke_dict = self.test_inside_coner(bmp_x2, bmp_y2, bmp_x3, bmp_y3, bmp_x4, bmp_y4, self.config.STROKE_MIN, inside_stroke_dict)
                        #print("inside_stroke_flag2:",inside_stroke_flag2)

                    if inside_stroke_flag1 or inside_stroke_flag2:
                        is_match_pattern = True

                if is_debug_mode:
                    if not is_match_pattern:
                        print(idx,"debug fail_code #3:", fail_code)
                    else:
                        print("match rule #3:",idx)

                if is_match_pattern:
                    #print("match rule #3")
                    #print(idx,": debug rule3:",format_dict_array[idx]['code'])

                    # to avoid same code apply twice.
                    nodes_length = len(format_dict_array)
                    generated_code = format_dict_array[(idx+0)%nodes_length]['code']
                    #print("generated_code#3:", generated_code)
                    apply_rule_log.append(generated_code)

                    if self.config.PROCESS_MODE in ["D"]:
                        generated_code = format_dict_array[(idx+1)%nodes_length]['code']
                        #print("generated_code#3+1:", generated_code)
                        apply_rule_log.append(generated_code)

                    is_goto_apply_round = True

                    # for D.Lucy
                    if self.config.PROCESS_MODE in ["D"]:
                        is_match_d_base_rule, fail_code = self.going_d_right(format_dict_array,idx)
                        is_goto_apply_round = is_match_d_base_rule

                    # for XD
                    if self.config.PROCESS_MODE in ["XD"]:
                        is_match_d_base_rule, fail_code = self.going_xd_down(format_dict_array,idx)
                        is_goto_apply_round = is_match_d_base_rule

                    # for RAINBOW
                    if self.config.PROCESS_MODE in ["RAINBOW"]:
                        is_match_d_base_rule, fail_code = self.going_rainbow_up(format_dict_array,idx)
                        is_goto_apply_round = is_match_d_base_rule

                    # NUT8, alway do nothing but record the history.
                    if self.config.PROCESS_MODE in ["NUT8"]:
                        is_goto_apply_round = False
                        generated_code = format_dict_array[(idx+1)%nodes_length]['code']
                        apply_rule_log.append(generated_code)

                    #print("is_goto_apply_round:",is_goto_apply_round)
                    if is_goto_apply_round:
                        center_x,center_y = -9999,-9999
                        #print("self.config.PROCESS_MODE:", self.config.PROCESS_MODE)
                        if not self.config.PROCESS_MODE in ["3TSANS"]:
                            center_x,center_y = self.apply_round_transform(format_dict_array,idx,apply_rule_log,generate_rule_log)
                        else:
                            center_x,center_y = self.apply_3t_transform(format_dict_array,idx,apply_rule_log,generate_rule_log)
                        #print("center_x,center_y:",center_x,center_y)

                    redo_travel=True
                    check_first_point = True
                    resume_idx = -1
                    break

        if check_first_point:
            # check close path.
            self.reset_first_point(format_dict_array, spline_dict)


        return redo_travel, resume_idx, inside_stroke_dict,apply_rule_log,generate_rule_log


python/TtfConfig.py
#!/usr/bin/env python3
#encoding=utf-8

class TtfConfig():
    VERSION = "2.099"
    PROCESS_MODE = "GOTHIC"
    #PROCESS_MODE = "HALFMOON"
    #PROCESS_MODE = "D"
    #PROCESS_MODE = "B2"
    #PROCESS_MODE = "B4"
    #PROCESS_MODE = "XD"
    #PROCESS_MODE = "RAINBOW"
    #PROCESS_MODE = "NUT8"
    #PROCESS_MODE = "3TSANS"
    #PROCESS_MODE = "TOOTHPASTE"
    PROCESS_MODE = "BAT"

    STYLE_INDEX = 5
    STYLE_ARRAY = ["Black","Bold","Medium","Regular","DemiLight","Light","Thin"]
    STYLE=STYLE_ARRAY[STYLE_INDEX]

    DEFAULT_COORDINATE_VALUE = -9999

    # for Regular
    STROKE_MAX = 102

    # 有些筆劃很細，有些很粗，設太細似乎又會誤判。
    # 雖然字變重，但因為筆畫複雜時，還是有細的線條。
    STROKE_MIN = 30

    # for X,Y axis equal compare.
    # each 100 px, +- 7 px.
    # PS: 大於等於8，其實滿可怕的。
    EQUAL_ACCURACY_MIN = 3
    EQUAL_ACCURACY_PERCENT = 0.07

    # for Regular
    ROUND_OFFSET = 33

    # don't access the distance within 5
    NEXT_DISTANCE_MIN = 5

    # for Regular
    OUTSIDE_ROUND_OFFSET = 55
    INSIDE_ROUND_OFFSET = 19

    NEED_LOAD_BMP_IMAGE = True

    if PROCESS_MODE in ["B2"]:
        NEED_LOAD_BMP_IMAGE = False
        INSIDE_ROUND_OFFSET = 30

        if STYLE=="Black":
            INSIDE_ROUND_OFFSET = 45
        if STYLE=="Bold":
            INSIDE_ROUND_OFFSET = 40
        if STYLE=="Medium":
            INSIDE_ROUND_OFFSET = 35

    if PROCESS_MODE in ["B4"]:
        NEED_LOAD_BMP_IMAGE = False
        INSIDE_ROUND_OFFSET = 25

        if STYLE=="Black":
            INSIDE_ROUND_OFFSET = 40
        if STYLE=="Bold":
            INSIDE_ROUND_OFFSET = 35
        if STYLE=="Medium":
            INSIDE_ROUND_OFFSET = 30

    # for Regular
    if PROCESS_MODE in ["TOOTHPASTE"]:
        OUTSIDE_ROUND_OFFSET = 65
        INSIDE_ROUND_OFFSET = 25

    # some inside block not able to fill 2 curve coner, use small one.
    INSIDE_SMALL_ROUND_OFFSET=15

    # unicode in field
    # 1 to 3
    UNICODE_FIELD = 2

    BMP_PATH = '/Users/chunyuyao/Documents/noto/bmp'


    def apply_weight_setting(self):
        self.STYLE=self.STYLE_ARRAY[self.STYLE_INDEX]

        # for uni53BB, 去, 155
        if self.STYLE=="Black":
            self.STROKE_MAX = 165
        if self.STYLE=="Bold":
            self.STROKE_MAX = 139
        if self.STYLE=="Medium":
            self.STROKE_MAX = 115
        if self.STYLE=="DemiLight":
            self.STROKE_MAX = 96
        if self.STYLE=="Light":
            self.STROKE_MAX = 80
        if self.STYLE=="Thin":
            self.STROKE_MAX = 59

        if self.STYLE=="Black":
            self.STROKE_MIN = 34
        if self.STYLE=="Bold":
            self.STROKE_MIN = 34
        if self.STYLE=="Medium":
            self.STROKE_MIN = 30
        if self.STYLE=="DemiLight":
            self.STROKE_MIN = 27
        if self.STYLE=="Light":
            self.STROKE_MIN = 26
        if self.STYLE=="Thin":
            self.STROKE_MIN = 24

        self.STROKE_ACCURACY_PERCENT = 10
        self.STROKE_WIDTH_MAX = int((self.STROKE_MAX * (100+self.STROKE_ACCURACY_PERCENT))/100)
        self.STROKE_WIDTH_MIN = int((self.STROKE_MIN * (100-self.STROKE_ACCURACY_PERCENT))/100)
        self.STROKE_WIDTH_AVERAGE = int((self.STROKE_MIN+self.STROKE_MAX)/2)
        #print("STROKE_WIDTH_MAX:", STROKE_WIDTH_MAX)
        #print("STROKE_MIN:", STROKE_WIDTH_MIN)


        if self.STYLE=="Black":
            self.ROUND_OFFSET = 48
        if self.STYLE=="Bold":
            self.ROUND_OFFSET = 41
        if self.STYLE=="Medium":
            self.ROUND_OFFSET = 37
        if self.STYLE=="DemiLight":
            self.ROUND_OFFSET = 30
        if self.STYLE=="Light":
            self.ROUND_OFFSET = 27
        if self.STYLE=="Thin":
            self.ROUND_OFFSET = 24

        if self.STYLE=="Black":
            self.OUTSIDE_ROUND_OFFSET = 65
        if self.STYLE=="Bold":
            self.OUTSIDE_ROUND_OFFSET = 60
        if self.STYLE=="Medium":
            self.OUTSIDE_ROUND_OFFSET = 60
        if self.STYLE=="DemiLight":
            self.OUTSIDE_ROUND_OFFSET = 50
        if self.STYLE=="Light":
            self.OUTSIDE_ROUND_OFFSET = 40
        if self.STYLE=="Thin":
            self.OUTSIDE_ROUND_OFFSET = 30

        # bigger curve.
        if self.PROCESS_MODE in ["TOOTHPASTE"]:
            if self.STYLE=="Black":
                self.OUTSIDE_ROUND_OFFSET = 75
            if self.STYLE=="Bold":
                self.OUTSIDE_ROUND_OFFSET = 70
            if self.STYLE=="Medium":
                self.OUTSIDE_ROUND_OFFSET = 65
            if self.STYLE=="DemiLight":
                self.OUTSIDE_ROUND_OFFSET = 60
            if self.STYLE=="Light":
                self.OUTSIDE_ROUND_OFFSET = 50
            if self.STYLE=="Thin":
                self.OUTSIDE_ROUND_OFFSET = 35

        # 只需要大彎.
        if self.PROCESS_MODE in ["D","XD","HALFMOON","NUT8","TOOTHPASTE"]:
            self.ROUND_OFFSET=self.OUTSIDE_ROUND_OFFSET

    def __init__(self, weight_code):
        import datetime

        self.STYLE_INDEX = int(weight_code)
        self.apply_weight_setting()
        print("Transform Mode:", self.PROCESS_MODE)
        print("Transform Style:", self.STYLE)
        print("Transform Version:", self.VERSION)
        print("Transform Time:", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))

    def hello(self):
        print("world!")